nav2_experimental/nav2_rl/nav2_turtlebot3_rl/random_crawl_action.py

```diff
@@ -0,0 +1,114 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os.path
+from ament_index_python.packages import get_package_share_directory
+from turtlebot3_env import TurtlebotEnv
+import numpy as np
+import rclpy
+import parameters
+
+from rclpy.node import Node
+from rclpy.action import ActionServer, CancelResponse, GoalResponse
+from time import sleep
+from keras.models import load_model
+
+import time
+
+from nav2_msgs.action import RandomCrawl
+from rclpy.executors import MultiThreadedExecutor
+from rclpy.callback_groups import ReentrantCallbackGroup
+import threading
+
+class RandomCrawlActionServer(Node):
+
+    def __init__(self):
+        super().__init__('action_server')
+        
+        self.env = TurtlebotEnv()
+        self.action_size = self.env.action_space()
+        print(self.action_size)
+        self.state = self.env.reset()
+        self.observation_space = len(self.state)
+        self.state = np.reshape(self.state, [1, self.observation_space])
+        pkg_share_directory = get_package_share_directory('nav2_turtlebot3_rl')
+        path = os.path.join(pkg_share_directory,"saved_models/random_crawl_waffle.h5")
+        self.model = load_model(path)
+        self.cancel_action = False
+        q = self.model.predict(self.state)
+        
+        self._goal_handle = None
+        self._goal_lock = threading.Lock()
+        self._action_server = ActionServer(
+            self,
+            RandomCrawl,
+            'RandomCrawl',
+            execute_callback=self.execute_callback,
+            handle_accepted_callback=self.handle_accepted_callback,
+            goal_callback=self.goal_callback,
+            cancel_callback=self.cancel_callback,
+            callback_group=ReentrantCallbackGroup())
+
+    def destroy(self):
+        self.env.cleanup()
+        self._action_server.destroy()
+        super().destroy_node()
+
+    def goal_callback(self, goal_request):
+        self.get_logger().info('Received goal request')
+        self.cancel_action = False
+        return GoalResponse.ACCEPT
+    
+    def handle_accepted_callback(self, goal_handle):
+        with self._goal_lock:
+            # This server only allows one goal at a time
+            if self._goal_handle is not None and self._goal_handle.is_active:
+                self.get_logger().info('Aborting previous goal')
+                # Abort the existing goal
+                self._goal_handle.abort()
+            self._goal_handle = goal_handle
+
+        goal_handle.execute()
+
+    def cancel_callback(self, goal_handle):
+        self.get_logger().info('Received cancel request')
+        self.cancel_action = True
+        return CancelResponse.ACCEPT
+
+    def execute_callback(self, goal_handle):
+        while not self.cancel_action and rclpy.ok():
```

Comment:
maybe remove `cancel_action` and add checks for `goal_handle.is_active` and `goal_handle.is_cancel_requested`